### PR TITLE
Add more invariant checks to ISRT

### PR DESCRIPTION
**Pull request by DaveCTurner**
Today there are two invariants of all realistic `IndexShardRoutingTable`
instances which are not enforced, nor are they satisfied by all instances
created by tests:

- every `IndexShardRoutingTable` must have exactly one primary shard
- if the primary shard is not active (`STARTED` or `RELOCATING`) then the
  replicas must all be `UNASSIGNED`

This commit adds assertions to enforce these invariants, and fixes up the tests
which create instances that do not satisfy them.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/routing/IndexShardRoutingTable.java [hunk 1]
```diff
@@ -559,6 +559,7 @@ public IndexShardRoutingTable build() {
             // don't allow more than one shard copy with same id to be allocated to same node
             assert distinctNodes(shards) : "more than one shard with same id assigned to same node (shards: " + shards + ")";
             assert noDuplicatePrimary(shards) : "expected but did not find unique primary in shard routing table: " + shards;
+            assert noAssignedReplicaWithoutActivePrimary(shards) : "unexpected assigned replica with no active primary: " + shards;
             return new IndexShardRoutingTable(shardId, shards);
```
[the remark this answers is not in the document]
Comment (idegtiarenko) [reply server/src/main/java/org/elasticsearch/cluster/routing/IndexShardRoutingTable.java:563]: Aha: https://github.com/elastic/elasticsearch/blob/00b3721108d749dcaf88ee187a61e66b21006060/server/src/main/java/org/elasticsearch/cluster/routing/IndexShardRoutingTable.java#L592-L593

[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/routing/IndexShardRoutingTable.java:563]: Yeah maybe I'll get to this at some point but today there are too many tests that violate this.

File: server/src/test/java/org/elasticsearch/cluster/routing/allocation/BalanceConfigurationTests.java [hunk 2]
```diff
@@ -318,155 +312,4 @@ public void testPersistedSettings() {
         assertThat(allocator.getShardBalance(), Matchers.equalTo(0.1f));
         assertThat(allocator.getThreshold(), Matchers.equalTo(3.0f));
     }
-
-    public void testNoRebalanceOnPrimaryOverload() {
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/cluster/routing/allocation/BalanceConfigurationTests.java:322]: I don't think this test was testing anything much. It sets up a specific routing table and then runs the balancer, but it only asserts that all the _assigned_ shards remain `STARTED` which is unsurprising and surely covered elsewhere. It seems to be expecting that the balancer would spread out the primaries but I don't think it actually does.

File: server/src/main/java/org/elasticsearch/cluster/routing/IndexShardRoutingTable.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/routing/IndexShardRoutingTable.java:563]: Actually it looks like this PR might fix all but the empty cases, see 58ca10a3bbd210678701b19e3767a0fea442b9e2.

[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/routing/IndexShardRoutingTable.java:563]: Moreover it looks like there's only a couple of empty cases too, see cda3a575bce62f868c4e75d2975e7fd38beeb58c.

**Review by idegtiarenko — APPROVED** [review]

**Comment by DaveCTurner** [comment]
@elasticmachine update branch

**Pull request merged**